### PR TITLE
Improve or-with disjoint checks

### PR DESCRIPTION
**Pull request by mvlabat**
# Objective

This PR attempts to improve query compatibility checks in scenarios involving `Or` filters.

Currently, for the following two disjoint queries, Bevy will throw a panic:

```
fn sys(_: Query<&mut C, Or<(With<A>, With<B>)>>, _: Query<&mut C, (Without<A>, Without<B>)>) {}
``` 

This PR addresses this particular scenario.

## Solution

`FilteredAccess::with` now stores a vector of `AccessFilters` (representing a pair of `with` and `without` bitsets), where each member represents an `Or` "variant".
Filters like `(With<A>, Or<(With<B>, Without<C>)>` are expected to be expanded into `A * B + A * !C`.

When calculating whether queries are compatible, every `AccessFilters` of a query is tested for incompatibility with every `AccessFilters` of another query.

---

## Changelog

- Improved system and query data access compatibility checks in scenarios involving `Or` filters

### CODE REVIEW

**Comment by alice-i-cecile** [comment]
Wording improvement for you.

> Improve compatibility checks in scenarios involving Or filters

Improved system and query data access compatibility checks in scenarios involving `Or` filters.

File: crates/bevy_ecs/src/query/access.rs [hunk 1]
```diff
@@ -325,6 +354,57 @@ impl<T: SparseSetIndex> FilteredAccess<T> {
     }
 }
 
+// A struct to express something like `Or<(With<A>, With<B>)>`.
+// Filters like `(With<A>, Or<(With<B>, With<C>)>` are expanded into `Or<(With<(A, B)>, With<(B, C)>)>`.
+#[derive(Clone, Eq, PartialEq)]
+struct ExpandedOrWithAccesses {
+    arr: SmallVec<[FixedBitSet; 8]>,
+}
```
Comment (mvlabat) [inline crates/bevy_ecs/src/query/access.rs:362]: I'm definitely open to renaming anything related to this struct, these are working names.
I'm also not sure that semantics of `extend`, `union` methods make sense now.

File: crates/bevy_ecs/src/query/access.rs [hunk 2]
```diff
@@ -289,7 +318,7 @@ impl<T: SparseSetIndex> FilteredAccess<T> {
 
     pub fn extend_intersect_filter(&mut self, other: &FilteredAccess<T>) {
         self.without.intersect_with(&other.without);
-        self.with.intersect_with(&other.with);
+        self.with.extend_with_or(&other.with);
     }
```
Comment (mvlabat) [inline crates/bevy_ecs/src/query/access.rs:322]: I'm kinda surprised that this is basically the only method I had to change to make tests work. Are there any other things that we have to consider?

Still not sure that the implementation is correct. I'll appreciate it if anyone can suggest other scenarios to cover with tests.

File: crates/bevy_ecs/src/query/access.rs [hunk 3]
```diff
@@ -325,6 +354,57 @@ impl<T: SparseSetIndex> FilteredAccess<T> {
...
 }
 
+// A struct to express something like `Or<(With<A>, With<B>)>`.
+// Filters like `(With<A>, Or<(With<B>, With<C>)>` are expanded into `Or<(With<(A, B)>, With<(B, C)>)>`.
+#[derive(Clone, Eq, PartialEq)]
+struct ExpandedOrWithAccesses {
+    arr: SmallVec<[FixedBitSet; 8]>,
```
Comment (mvlabat) [inline crates/bevy_ecs/src/query/access.rs:361]: I used `SmallVec` to avoid introducing another level of indirection. Is it worth it? Does the inline capacity make sense?

File: crates/bevy_ecs/src/query/access.rs [hunk 4]
```diff
@@ -325,6 +354,57 @@ impl<T: SparseSetIndex> FilteredAccess<T> {
     }
 }
 
+// A struct to express something like `Or<(With<A>, With<B>)>`.
+// Filters like `(With<A>, Or<(With<B>, With<C>)>` are expanded into `Or<(With<(A, B)>, With<(B, C)>)>`.
```
Comment (MinerSebas) [inline crates/bevy_ecs/src/query/access.rs:358]: The proper syntax for `and`ed `With`es is a tuple around the `With` and not inside. (Also fixes the wrong `With<(B, C)`)
```suggestion
// Filters like `(With<A>, Or<(With<B>, With<C>)>` are expanded into `Or<((With<A>, With<B>), (With<A>, With<C>))>`.
```

Comment (mvlabat) [reply crates/bevy_ecs/src/query/access.rs:358]: Ah, good catch, thanks

File: crates/bevy_ecs/src/query/access.rs [hunk 5]
```diff
@@ -41,6 +44,28 @@ impl<'a, T: SparseSetIndex + fmt::Debug> fmt::Debug for FormattedBitSet<'a, T> {
...
+    }
+}
+
+impl<'a, T: SparseSetIndex + fmt::Debug> fmt::Debug for FormattedExpandedOrWithAccesses<'a, T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_list()
+            .entries(self.with.arr.iter().map(FormattedBitSet::<T>::new))
+            .finish()
+    }
+}
```
Comment (MinerSebas) [inline crates/bevy_ecs/src/query/access.rs:67]: Why create a Proxy type here, instead of implementing Debug for ExpandedOrWithAccesses?
The `FormattedBitSet` is used because the `FixedBitSet` it wraps is a foreign type with a separate debug impl.

**Comment by infmagic2047** [comment]
The implementation looks correct to me. The methods definitely need better names, now that the filter is no longer represented as a set (maybe names based on "and/or").

We can use the same `ExpandedOrWithAccess` for `Without` filters with very little changes (I think only `is_disjoint` needs to be changed), so it may be a good idea.

**Comment by alice-i-cecile** [comment]
> The implementation looks correct to me. The methods definitely need better names, now that the filter is no longer represented as a set (maybe names based on "and/or").
> 
> We can use the same `ExpandedOrWithAccess` for `Without` filters with very little changes (I think only `is_disjoint` needs to be changed), so it may be a good idea.

I would like to see both of these things done in this PR.

File: crates/bevy_ecs/src/query/access.rs (hunk 3 above)
Comment (rodolphito) [reply crates/bevy_ecs/src/query/access.rs:361]: This can only be answered with benchmarks, have you run before/afters? I can try on my machine.

**Review by rodolphito — APPROVED** [review]
This mostly makes sense, I don't see where union_with is used though.

File: crates/bevy_ecs/src/query/access.rs [hunk 6]
```diff
@@ -325,6 +354,57 @@ impl<T: SparseSetIndex> FilteredAccess<T> {
     }
 }
 
+// A struct to express something like `Or<(With<A>, With<B>)>`.
+// Filters like `(With<A>, Or<(With<B>, With<C>)>` are expanded into `Or<((With<A>, With<B>), (With<A>, With<C>))>`.
+#[derive(Clone, Eq, PartialEq)]
+struct ExpandedOrWithAccesses {
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/query/access.rs:360]: I think this would be clearer if it was just `WithAccesses`, and explained further about `Or` behavior in the doc strings.

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
I'd like to see the small comments resolved, and add another system compatibility test for nested Ors:

```rust
        fn sys(_: Query<&mut D, Or<(Or<(With<A>, With<B>)>), (Or<(With<A>, With<C>))>>>, _: Query<&mut D, Without<A>>) {}
```
I have zero confidence that I got those brackets right, but you get the idea.

File: crates/bevy_ecs/src/query/access.rs (hunk 3 above)
Comment (mvlabat) [reply crates/bevy_ecs/src/query/access.rs:325]: Yeah, I'd be grateful for the help with benchmarking these. I haven't run any benchmarks myself.

File: crates/bevy_ecs/src/query/access.rs (hunk 5 above)
Comment (mvlabat) [reply crates/bevy_ecs/src/query/access.rs]: Yeah, I copy-pasted the approach for `FixedBitSet ` without giving it much thought.. You're right, it's not needed. I had to introduce the index marker type for the debug implementation though.

**Comment by mvlabat** [comment]
Updated the implementation to support `Or` filters for `Without`, added more tests (including the one suggested by Alice).

File: crates/bevy_ecs/src/query/access.rs [hunk 7]
```diff
@@ -325,6 +318,84 @@ impl<T: SparseSetIndex> FilteredAccess<T> {
     }
 }
 
+// A struct to express something like `Or<(With<A>, With<B>)>`.
+// Filters like `(With<A>, Or<(With<B>, With<C>)>` are expanded into `Or<((With<A>, With<B>), (With<A>, With<C>))>`.
+#[derive(Clone, Eq, PartialEq)]
+struct ExpandedOrAccess<T, I> {
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/query/access.rs:324]: I think just `ExpandedAccess` or `NormalizedAccess` is clearer and easier to generalize later.

File: crates/bevy_ecs/src/query/access.rs [hunk 8]
```diff
@@ -325,6 +318,84 @@ impl<T: SparseSetIndex> FilteredAccess<T> {
     }
 }
 
+// A struct to express something like `Or<(With<A>, With<B>)>`.
+// Filters like `(With<A>, Or<(With<B>, With<C>)>` are expanded into `Or<((With<A>, With<B>), (With<A>, With<C>))>`.
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/query/access.rs:322]: I'd like to see better docs explaning the roles of the generics. It's mentioned below, but it can be hard to follow when you're encountering this type for the first time.

File: crates/bevy_ecs/src/query/access.rs (hunk 3 above)
Comment (alice-i-cecile) [reply crates/bevy_ecs/src/query/access.rs:325]: Unless we have benchmarks I'd prefer to just use a Vec here for clarity.

File: crates/bevy_ecs/src/query/access.rs [hunk 9]
```diff
@@ -325,6 +318,84 @@ impl<T: SparseSetIndex> FilteredAccess<T> {
...
+#[derive(Clone, Eq, PartialEq)]
+struct ExpandedOrAccess<T, I> {
+    arr: SmallVec<[FixedBitSet; 8]>,
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/query/access.rs:325]: I'd like to see more docs on exactly how this data structure works. Why do we have multiple fixed bitsets? How does it interact?

File: crates/bevy_ecs/src/query/access.rs [hunk 10]
```diff
@@ -325,6 +318,84 @@ impl<T: SparseSetIndex> FilteredAccess<T> {
...
+    fn default() -> Self {
+        Self {
+            arr: smallvec::smallvec![FixedBitSet::default()],
+            _filter_type: PhantomData,
+            _index_type: PhantomData,
+        }
+    }
+}
+
+impl<T, I: SparseSetIndex> ExpandedOrAccess<T, I> {
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/query/access.rs:355]: Each of these methods needs to be carefully commented. This is tricky soundness critical code, and I'd like to see the logic for each more explicitly.

File: crates/bevy_ecs/src/query/access.rs [hunk 11]
```diff
@@ -325,6 +318,84 @@ impl<T: SparseSetIndex> FilteredAccess<T> {
...
+                new_with.push(w);
+            }
+        }
+        self.arr = new_with;
+    }
+}
+
+impl<I: SparseSetIndex> ExpandedOrAccess<WithMarker, I> {
+    fn is_disjoint(&self, other: &ExpandedOrAccess<WithoutMarker, I>) -> bool {
+        // We need to make sure that every `without` variant is disjoint with any `with`.
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/query/access.rs:389]: This statement isn't obviously correct: can you add more justification or otherwise improve clarity?

File: crates/bevy_ecs/src/query/access.rs [hunk 12]
```diff
@@ -532,4 +606,38 @@ mod tests {
 
         assert!(access_a.eq(&expected));
     }
+
+    #[test]
+    fn filtered_access_extend_or() {
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/query/access.rs:611]: It's hard for me to understand the exact invariants that this test is upholding. I think that this test either needs more comments, to be split apart, or to be cut due to being tightly coupled to the internal API.

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
Being unusually harsh around clarity here: this code was already tricky, and I'm nervous about ongoing correctness and ability to change this code as needed.

File: crates/bevy_ecs/src/query/access.rs [hunk 13]
```diff
@@ -325,6 +318,84 @@ impl<T: SparseSetIndex> FilteredAccess<T> {
...
+            }
+        }
+        self.arr = new_with;
+    }
+}
+
+impl<I: SparseSetIndex> ExpandedOrAccess<WithMarker, I> {
+    fn is_disjoint(&self, other: &ExpandedOrAccess<WithoutMarker, I>) -> bool {
+        // We need to make sure that every `without` variant is disjoint with any `with`.
+        // For example, `Or<(Without<A>, Without<B>)>` is disjoint with `Or<(With<A>, With<B>)>`,
```
Comment (SkiFire13) [inline crates/bevy_ecs/src/query/access.rs:390]: This doesn't looks correct to me. Both `Or<(Without<A>, Without<B>)>` and `Or<(With<A>, With<B>)>` match entities that have exactly one component among `A` and `B` but not both, so they're not completly disjoint.

**Comment by SkiFire13** [comment]
The short discussion on Discord prompted me to try to explain what this PR does as operations on boolean SAT formulas. I feel like this might be a nice addition to the internal documentation:

The new `ExpandedOrAccess` expresses the with/without clauses as a boolean SAT formula in disjunctive normal form (i.e. a sequence of OR of formulas that contain a sequence of AND of either terms or NOT of terms).
In the query type, normal tuples represent AND formulas while `Or<(...)>` represent OR formulas. `With<T>` represent a normal term while `Without<T>` represents a NOT term.

The sequence of ORs is represented as two `SmallVec<[FixedBitSet; 8]>` (one in each `ExpandedOrAccess`) and the inner formulas with AND is split in two `FixedBitSet`s, one in each `ExpandedOrAccess`. One contains the terms that are "positive" (in the `with` `ExpandedOrAccess`) and the other the terms that are negative, i.e. are behind a NOT (in the `without` `ExpandedOrAccess`).
(Observation: wouldn't it be more intuitive to place the `with` and `without` `FixedBitSet` in the same structure, and then place instances of that one in a single `SmallVec<[NewStructure; 8]>`, in order to be more similar to the AND formulas?)

Regarding the operations:
`extend_intersect_filter` is essentially an OR between two sequences of ORs, so they can just be chained together.

`add_with` is essentially computing `A & ((B1 & ... & Bn) | (C1 & ... & Cn) | ...)`, which is equivalent to `(B1 & ... & Bn & A) | (C1 & ... & Cn & A) | ...`, that is we need to insert `A` into each existing AND formula (i.e. into each `FixedBitSet` in the `with` or `without` `ExpandedOrAccess` depends on whether `A` is positive (`With<T>`) or negative (`Without<T>`))

`extend`/`union_with` is essentially computing `((A1 & ... & An) | (B1 & ... & Bn) | ...) & ((C1 & ... & Cn) | (D1 & ... & Dn) | ...)`, which is equivalent to `(A1 & ... & An & C1 & ... & Cn) | (A1 & ... & An & D1 & ... & Dn) | (B1 & ... & Bn & C1 & ... & Cn) | (B1 & ... & Bn & D1 & ... & Dn) | ...`, that is it computes the cartesian product of the two lists of AND formulas and maps each "tuple" to an AND formula, which can be flattened.

`is_disjoint` is equivalent to computing `extend`/`union_with` (with the difference that it can now be computed lazily) and then checking if the resulting formula can ever be satisfied, that is if any of the resulting AND formulas can ever be satisfied. Thus this reduces to check for each tuple of AND formula if they are disjoint, which can be computed by checking that both `with` `FixedBitSet` are disjoint with each `without` `FixedBitSet`. The PR does this in `Filteredaccess::is_compatible` and then `ExpandedOrAccess::is_disjoint`.
(Observation: the PR only checks two combinations of `with` and `without`, thus it assumes that a single `FilteredAccess` doesn't contain conflicting filters)

**Comment by mvlabat** [comment]
@SkiFire13 thanks for the excellent write-up! It's quite on point, and it's even been educational for me: for example, I was looking myself for terminology such as DNF to describe the idea behind this. You've done a much better job, as I still struggle with correct semantics, naming and documenting things :D

> (Observation: wouldn't it be more intuitive to place the with and without FixedBitSet in the same structure, and then place instances of that one in a single SmallVec<[NewStructure; 8]>, in order to be more similar to the AND formulas?)

I was considering it, but I found it made the code much messier. Lengths of `with` and `without` arrays can be different, which requires additional workarounds to combine them into a single array, introducing more problems than it would actually solve.

> (Observation: the PR only checks two combinations of with and without, thus it assumes that a single FilteredAccess doesn't contain conflicting filters)

I'm afraid I didn't understand this part, could you expand on that, please? I'm trying to understand if that means that I overlooked some edge cases.

File: crates/bevy_ecs/src/query/access.rs (hunk 13 above)
Comment (mvlabat) [reply crates/bevy_ecs/src/query/access.rs:390]: Yeah, I think you're right, that doesn't fully make sense. "Disjoint" is probably a poor word choice, but what I meant is that for the compatibility check it matters only if every `without` AND sequence is disjoint with at least one `with` AND sequence (in which case we can't guarantee that the queries are compatible).

(I'm trying to argue that the comment is correct in the context of how the method is used, apart from the fact that "disjoint" is a bad term here.)

Will it help if we rename the method to something like `is_ruled_out_by` and negate the returned result? Then the semantics would be
```rust
pub fn is_compatible(&self, other: &FilteredAccess<T>) -> bool {
    self.access.is_compatible(&other.access)
        || self.with.is_ruled_out_by(&other.without)
        || other.with.is_ruled_out_by(&self.without)
}
```

I hope it'll help to paraphrase the comment to be less confusing.

**Comment by mvlabat** [comment]
@alice-i-cecile thank you for the comments! That's absolutely fair, I totally agree that things can be named and documented better. I feel somewhat limited by my vocabulary and discrete maths knowledge, so I'll appreciate direct suggestions for naming things and documentation. I feel a bit awkward asking others to document my code, so I'll do my best to improve the bits that I can, but I suspect that this PR will need an additional round of polish from someone with better tech writing skills ☺️

File: crates/bevy_ecs/src/query/access.rs (hunk 7 above)
Comment (mvlabat) [reply crates/bevy_ecs/src/query/access.rs]: Thanks, I went with `NormalizedAccess`

File: crates/bevy_ecs/src/query/access.rs (hunk 8 above)
Comment (mvlabat) [reply crates/bevy_ecs/src/query/access.rs:322]: I've expanded the docs a bit, let me know if that works

**Comment by SkiFire13** [comment]
> I was considering it, but I found it made the code much messier. Lengths of `with` and `without` arrays can be different

I'm not sure how that can happen. Looks like every operation that extends one of them is immediately executed on the other one too, preserving the length equatility.

> I'm afraid I didn't understand this part, could you expand on that, please? I'm trying to understand if that means that I overlooked some edge cases.

If you have a filter like `(With<T>, Without<T>)` then this can never match and thus is compatible with every query. In boolean logic this translates to every `AND` sequence containing both a `T` and a `NOT T` term. This can be checked by checking if `self.with` is ruled out by `self.without` (and equivalently if `other.with` is ruled out by `other.without`, essentially checking all the `self` and `other` combinations with `is_ruled_out_by`.

---

One additional observation: checking `self.with.is_ruled_out_by(other.without)` and `other.with.is_ruled_out_by(self.without)` separately is not enough. Consider for example `Or<(With<A>, Without<B>)>` and `(With<B>, Without<A>)`. These two filters are actually disjoint, but if you manually calculate their `with` and `without` they are:

| |`with`|`without`|
|-|-|-|
|`f1`: `Or<(With<A>, Without<B>)>`|`[[A], []]`|`[[], [B]]`|
|`f2`: `(With<B>, Without<A>)`|`[[B]]`|`[[A]]`|

When checking whether `f1.with` is ruled out by `f2` you can see that `[]` and `[A]` are disjoint, so it's not ruled out.
However when checking whether `f2.with` is ruled out by `f1.without` you can see that `[]` and `[B]` are disjoint again, so it's not ruled out. So according to the algorithm they may conflict, but in practice we can reason that they can't (in other for an entity to match `f1` it has to either have an `A`, but then it would not match the `Without<A>` in `f2`, or not have a `B`, but then it would not match the `With<B>` in `f2`).

So what's the error? We're requiring all of `f1`'s `with`s to be ruled out by `f2`'s `without` or the opposite, but a mix of them could happen: sometimes one rules out the other and sometimes the opposite, but one of them always happen for every combination of `AND` formula from the two filters. It becomes clear when you zip the `with` and `without`s, like I previously observed. Then you get:

`Or<(With<A>, Without<B>)>`: `[ { with: [A], without: [] }, { with: [], without: [B] } ]`
`(With<B>, Without<A>)`: `[ { with: [B], without: [A] } ]`

Finally, looking at every combination to see if they are compatible:

- `{ with: [A], without: [] }` and `{ with: [B], without: [A] }`
  - the first `with` is ruled out by the second `without`, so they are compatible
- `{ with: [], without: [B] }` and `{ with: [B], without: [A] }`
  - the second's `with` is ruled out by the first `with`, so they are compatible

File: crates/bevy_ecs/src/query/access.rs (hunk 13 above)
Comment (SkiFire13) [reply crates/bevy_ecs/src/query/access.rs:393]: > Will it help if we rename the method to something like `is_ruled_out_by` and negate the returned result?

`is_ruled_out_by` looks like a better term to me. Disjoint is definitely confusing because the `with`/`without` AND sequences being disjoint mean the opposite of the queries being disjoint.

> it matters only if every `without` AND sequence is disjoint with at least one `with` AND sequence

This seems flawed to me. What if not every `without` AND sequence is disjoint with at least one `with` AND sequence, but at least one is? Then you still get one set of possible, which means the two queries are not actually compatible.

Actually though I think looking at them separately is wrong. See my comment below.

File: crates/bevy_ecs/src/query/access.rs (hunk 12 above)
Comment (mvlabat) [reply crates/bevy_ecs/src/query/access.rs:627]: Added more comments to the test, I hope it makes more sense now as a whole

**Comment by mvlabat** [comment]
@SkiFire13 I really appreciate all the effort you put into the discussion! I'll need to re-read your comment tomorrow morning and try to write a more complete response. But the thing I stumbled upon and I wanted to give a quick response is the following statement:

> `f1`: `Or<(With<A>, Without<B>)>` | `[[A], []]` | `[[], [B]]`

I don't think the code will actually produce this kind of data structure.
I'd expect it to result in `with: [[A]]` and `without: [[B]]`.

In this case, I believe the `is_ruled_out_by` check would behave correctly.

File: crates/bevy_ecs/src/system/mod.rs [hunk 14]
```diff
@@ -402,6 +409,82 @@ mod tests {
...
+        run_system(&mut world, sys);
+    }
+
+    #[test]
+    fn or_expanded_nested_with_and_common_nested_without() {
+        fn sys(
+            _: Query<&mut D, Or<((With<A>, With<B>), (With<B>, With<C>))>>,
+            _: Query<&mut D, Or<(Without<A>, Without<B>)>>,
+        ) {
+        }
```
Comment (SkiFire13) [inline crates/bevy_ecs/src/system/mod.rs:471]: This test shoult not pass: an entity could have components `B`, `C` and `D` but not `A`. In that cause it would match both the first query (due to the `(With<B>, With<C>)`) and the second query (due to the `Without<A>`).

Comment (mvlabat) [reply crates/bevy_ecs/src/system/mod.rs:471]: Great catch, thanks! I think I meant to write something different in the test, but it turns out that this wrong version proves that the logic is flawed.

**Comment by mvlabat** [comment]
@SkiFire13 huge thanks! Your comments were absolutely on point, the logic was indeed flawed, and combining `with` and `without` filters under one array helped to fix it. My first impression of this approach was wrong, it actually made the implementation simpler and easier to follow.

I've added tests with all the scenarios that you described where the previous approach was falling and confirmed that the algorithm now handles them correctly.

I really appreciate the effort you've put into reviewing the code and the explanations that you've provided.

File: crates/bevy_ecs/src/query/access.rs [hunk 15]
```diff
@@ -266,30 +261,46 @@ impl<T: SparseSetIndex> FilteredAccess<T> {
...
     }
 
-    /// Retains only combinations where the element given by `index` is also present.
-    pub fn add_with(&mut self, index: T) {
-        self.with.grow(index.sparse_set_index() + 1);
-        self.with.insert(index.sparse_set_index());
+    /// Adds a `With` filter: corresponds to a conjunction (AND) operation.
+    ///
+    /// For example, in case we have `Or<(With<A>, With<B>)>`, which is represented by an array of two `AccessFilter` instances,
+    /// adding `AND With<C>` gets expanded into `Or<((With<A>, With<C>), (With<B>, With<C>))>`.
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/query/access.rs:276]:
```suggestion
    /// Suppose we begin with `Or<(With<A>, With<B>)>`, which is represented by an array of two `AccessFilter` instances.
    /// Adding `AND With<C>` via this method transforms it into the equivalent of  `Or<((With<A>, With<C>), (With<B>, With<C>))>`.
```

File: crates/bevy_ecs/src/query/access.rs [hunk 16]
```diff
@@ -266,30 +261,46 @@ impl<T: SparseSetIndex> FilteredAccess<T> {
...
+        }
     }
 
-    /// Retains only combinations where the element given by `index` is not present.
-    pub fn add_without(&mut self, index: T) {
-        self.without.grow(index.sparse_set_index() + 1);
-        self.without.insert(index.sparse_set_index());
+    /// Adds a `Without` filter: corresponds to a conjunction (AND) operation.
+    ///
+    /// For example, in case we have `Or<(With<A>, With<B>)>`, which is represented by an array of two `AccessFilter` instances,
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/query/access.rs:287]: Same clarity / grammar suggestion as above.

File: crates/bevy_ecs/src/query/access.rs [hunk 17]
```diff
@@ -298,9 +309,23 @@ impl<T: SparseSetIndex> FilteredAccess<T> {
...
-        self.access.is_compatible(&other.access)
-            || !self.with.is_disjoint(&other.without)
-            || !other.with.is_disjoint(&self.without)
+        if self.access.is_compatible(&other.access) {
+            return true;
+        }
+
+        // If the access instances are incompatible, we want to check that whether filters can
+        // guarantee that queries are disjoint.
+        // Since the `filter_sets` array represents a DNF formula ("ORs of ANDs"),
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/query/access.rs:318]:
```suggestion
        // Since the `filter_sets` array represents a Disjunctive Normal Form formula ("ORs of ANDs"),
```

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
Dramatically better now that we have more tests, more docs, and a clear theoretical framework. Some clarity suggestions for the docs.

File: crates/bevy_ecs/src/query/access.rs (hunk 11 above)
Comment (mvlabat) [reply crates/bevy_ecs/src/query/access.rs]: (marking as resolved as the implementation and comments changed significantly)

File: crates/bevy_ecs/src/query/access.rs [hunk 18]
```diff
@@ -325,6 +375,38 @@ impl<T: SparseSetIndex> FilteredAccess<T> {
...
+            with: FixedBitSet::default(),
+            without: FixedBitSet::default(),
+            _index_type: PhantomData,
+        }
+    }
+}
+
+impl<T: SparseSetIndex> AccessFilters<T> {
+    fn is_ruled_out_by(&self, other: &Self) -> bool {
+        !self.with.is_disjoint(&other.without) || !self.without.is_disjoint(&other.with)
```
Comment (SkiFire13) [inline crates/bevy_ecs/src/query/access.rs:406]: Technically speaking to be complete this also needs `|| !self.with.is_disjoing(&self.without) || !other.with.is_disjoint(&other.without)` in order to handle extreme cases like `(With<T>, Without<T>)` being empty and thus disjoint with everything else. Not sure if we want to properly support those cases though, in practice if they happen they're almost always a mistake. The condition being at the end means it would be checked only if we're about to find a conflict, so the performance hit should not happen in the happy path.

However if it's kept like this I would add a comment explaining the missing conditions and why they're not implemented.

File: crates/bevy_ecs/src/query/fetch.rs [hunk 19]
```diff
@@ -1527,7 +1527,7 @@ macro_rules! impl_anytuple_fetch {
                     if _not_first {
```
Comment (SkiFire13) [inline crates/bevy_ecs/src/query/fetch.rs:1527]: Github doesn't let me select the lines above, but there's a comment claiming that the `with`/`without` access should be added only if all the params have that access, but this is no longer true, instead disjoint access is properly tracked with `append_or`.

File: crates/bevy_ecs/src/query/filter.rs [hunk 20]
```diff
@@ -366,7 +366,7 @@ macro_rules! impl_query_filter_tuple {
                     if _not_first {
```
Comment (SkiFire13) [inline crates/bevy_ecs/src/query/filter.rs:366]: Same here.

File: crates/bevy_ecs/src/query/access.rs (hunk 12 above)
Comment (alice-i-cecile) [reply crates/bevy_ecs/src/query/access.rs:622]: These are much better now, thank you.

File: crates/bevy_ecs/src/query/fetch.rs (hunk 19 above)
Comment (mvlabat) [reply crates/bevy_ecs/src/query/fetch.rs:1515]: Removed the comment and renamed variables a bit to avoid confusion with the "intersect" word

File: crates/bevy_ecs/src/query/filter.rs (hunk 20 above)
Comment (mvlabat) [reply crates/bevy_ecs/src/query/filter.rs:354]: Removed the comment and renamed variables a bit to avoid confusion with the "intersect" word

File: crates/bevy_ecs/src/query/access.rs (hunk 18 above)
Comment (mvlabat) [reply crates/bevy_ecs/src/query/access.rs:411]: Yeah, I was also thinking about this case. I'm inclined not to support it, I agree that it almost always means there's an error. Added a comment.

File: crates/bevy_ecs/src/query/filter.rs [hunk 21]
```diff
@@ -348,33 +348,21 @@ macro_rules! impl_query_filter_tuple {
...
+                let mut _new_access = access.clone();
                 let mut _not_first = false;
                 $(
                     if _not_first {
                         let mut intermediate = access.clone();
                         $filter::update_component_access($filter, &mut intermediate);
-                        _intersected_access.extend_intersect_filter(&intermediate);
-                        _intersected_access.extend_access(&intermediate);
+                        _new_access.append_or(&intermediate);
+                        _new_access.extend_access(&intermediate);
```
Comment (mvlabat) [inline crates/bevy_ecs/src/query/filter.rs:358]: This made me wonder if this line is actually needed. I see why we call `extend_access` for `AnyOf`, but I don't think we need to do this for `Or`. If anyone can confirm, I'll drop it.

File: crates/bevy_ecs/src/query/access.rs (hunk 9 above)
Comment (mvlabat) [reply crates/bevy_ecs/src/query/access.rs]: (the implementation evolved a bit, and so did the comments, resolving it in favour of comments to the up-to-date diff)

File: crates/bevy_ecs/src/query/access.rs (hunk 6 above)
Comment (mvlabat) [reply crates/bevy_ecs/src/query/access.rs]: (the implementation evolved a bit, and so did the comments, resolving it in favour of comments to the up-to-date diff)

File: crates/bevy_ecs/src/query/access.rs (hunk 8 above)
Comment (mvlabat) [reply crates/bevy_ecs/src/query/access.rs]: got rid of the additional generic argument eventually

File: crates/bevy_ecs/src/query/access.rs (hunk 10 above)
Comment (mvlabat) [reply crates/bevy_ecs/src/query/access.rs]: (the implementation evolved a bit, and so did the comments, resolving it in favour of comments to the up-to-date diff)

File: crates/bevy_ecs/src/query/access.rs (hunk 13 above)
Comment (mvlabat) [reply crates/bevy_ecs/src/query/access.rs]: Fixed and added the tests covering the previously failing scenarios

File: crates/bevy_ecs/src/query/access.rs [hunk 22]
```diff
@@ -213,31 +217,22 @@ impl<T: SparseSetIndex> Access<T> {
...
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        f.debug_struct("FilteredAccess")
-            .field("access", &self.access)
-            .field("with", &FormattedBitSet::<T>::new(&self.with))
-            .field("without", &FormattedBitSet::<T>::new(&self.without))
-            .finish()
-    }
+    // An array of filter sets to express `With` or `Without` clauses in disjunctive normal form, for example: `Or<(With<A>, With<B>)>`.
+    // Filters like `(With<A>, Or<(With<B>, Without<C>)>` are expanded into `Or<((With<A>, With<B>), (With<A>, Without<C>))>`.
+    filter_sets: SmallVec<[AccessFilters<T>; 8]>,
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/query/access.rs:228]: I'd prefer to remove the `SmallVec` here. It adds complexity in an area that's already overloaded and isn't supported by measurements.

We can add it back later if justified.

**Review by SkiFire13 — APPROVED** [review]

File: crates/bevy_ecs/src/query/access.rs (hunk 22 above)
Comment (mvlabat) [reply crates/bevy_ecs/src/query/access.rs]: Done 👍

**Review by alice-i-cecile — APPROVED** [review]
